dask/dataframe/io/io.py

```diff
@@ -454,48 +454,59 @@ def from_dask_array(x, columns=None, index=None, meta=None):
     """
     meta = _meta_from_array(x, columns, index, meta=meta)
 
-    if x.ndim == 2 and len(x.chunks[1]) > 1:
-        x = x.rechunk({1: x.shape[1]})
+    name = "from-dask-array-" + tokenize(x, columns)
+    graph_dependencies = [x]
+    arrays_and_indices = [x.name, "ij" if x.ndim == 2 else "i"]
+    numblocks = {x.name: x.numblocks}
 
-    name = "from-dask-array" + tokenize(x, columns)
-    to_merge = []
+    if is_series_like(meta):
+        kwargs = {"dtype": x.dtype, "name": meta.name}
+    else:
+        kwargs = {"columns": meta.columns}
 
     if index is not None:
-        if not isinstance(index, Index):
-            raise ValueError("'index' must be an instance of dask.dataframe.Index")
         if index.npartitions != x.numblocks[0]:
             msg = (
                 "The index and array have different numbers of blocks. "
                 "({} != {})".format(index.npartitions, x.numblocks[0])
             )
             raise ValueError(msg)
         divisions = index.divisions
-        to_merge.append(ensure_dict(index.dask))
-        index = index.__dask_keys__()
+        graph_dependencies.append(index)
+        arrays_and_indices.extend([index.__dask_layers__()[0], "i"])
+        numblocks[index.__dask_layers__()[0]] = (index.npartitions,)
 
     elif np.isnan(sum(x.shape)):
         divisions = [None] * (len(x.chunks[0]) + 1)
-        index = [None] * len(x.chunks[0])
     else:
+        dsk = {}
         divisions = [0]
-        for c in x.chunks[0]:
-            divisions.append(divisions[-1] + c)
-        index = [
-            (np.arange, a, b, 1, "i8") for a, b in zip(divisions[:-1], divisions[1:])
-        ]
+        stop = 0
+        index_name = name + "-index-" + tokenize(x.chunks[0])
+        for i, increment in enumerate(x.chunks[0]):
+            stop += increment
+            dsk[(index_name, i)] = (pd.RangeIndex, divisions[i], stop)
+            divisions.append(stop)
         divisions[-1] -= 1
+        index = new_dd_object(dsk, index_name, pd.RangeIndex(0, 1), divisions)
+        arrays_and_indices.extend([index.__dask_layers__()[0], "i"])
+        graph_dependencies.append(index)
+        numblocks[index.__dask_layers__()[0]] = (index.npartitions,)
+
+    blk = blockwise(
+        type(meta),
```

Comment:
```suggestion
        index_mapping = {}
        for i, increment in enumerate(x.chunks[0]):
            stop += increment
            index_mapping[(i,)] = (divisions[i], stop)
            divisions.append(stop)
        divisions[-1] -= 1
        arrays_and_indices.extend(
            [BlockwiseDepDict(mapping=index_mapping), "i"]
        )

    def _partition_from_array(data, index=None, **kwargs):
        if isinstance(index, tuple):
            index = pd.RangeIndex(*index)
        return type(meta)(data, index=index, **kwargs)

    blk = blockwise(
        _partition_from_array,
```

You will also need to add `BlockwiseDepDict` to the `dask.blockwise` imports